// @gt-npm/gt-lib-ts/es/h5GameTools/commonStore/index.mjs
import { useGlobalEventDispatcher } from '../../eventDispatcher/index.mjs';
import { insertVariablePattern } from '../../utils/objectUtils.mjs';
import '../../utils/reactivity/index.mjs';
import { SlotGameEvent } from '../slotGameConnector.mjs';
import { useGetters } from './getters.mjs';
import { useMutations } from './mutations.mjs';
import { useStoreState } from './state.mjs';
export { StoreState } from './state.mjs';
import { readonly } from '../../utils/reactivity/vue-reactivity.mjs';

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
class CommonStore {
  constructor() {
    __publicField(this, "_storeState", useStoreState());
    __publicField(this, "storeState", readonly(this._storeState));
    __publicField(this, "storeMutation", useMutations(this._storeState));
    __publicField(this, "storeGetter", useGetters(this._storeState));
    __publicField(this, "i18n");
    this.i18n = new Proxy({}, {
      get(_, prop) {
        return CommonStore.shared.storeState.i18n[prop] || prop;
      }
    });
  }
  static get shared() {
    var _a, _b;
    if ((_a = window.gtLibTsScope) == null ? void 0 : _a.commonStore) {
      return window.gtLibTsScope.commonStore;
    }
    return ((_b = window.gtLibTsScope) != null ? _b : window.gtLibTsScope = {}).commonStore = new CommonStore();
  }
  addListeners() {
    [
      [SlotGameEvent.READY, this.storeMutation.initState],
      [SlotGameEvent.LOGIN, this.storeMutation.onLogin],
      [SlotGameEvent.GET_MACHINE_LIST, this.storeMutation.initState],
      [SlotGameEvent.TAKE_MACHINE, this.storeMutation.onTakeMachine],
      [SlotGameEvent.LOAD_INFO, this.storeMutation.onOnLoadInfo],
      [SlotGameEvent.GET_MACHINE_DETAIL, this.storeMutation.onGetMachineDetail],
      [SlotGameEvent.CREDIT_EXCHANGE, this.storeMutation.onCreditExchange],
      [SlotGameEvent.BALANCE_EXCHANGE, this.storeMutation.onBalanceExchange],
      [SlotGameEvent.UPDATE_JP, this.storeMutation.initState],
      [SlotGameEvent.UPDATE_MARQUEE, this.storeMutation.initState],
      [SlotGameEvent.HIT_JACKPOT, this.storeMutation.onHitJackpot],
      [SlotGameEvent.END_GAME, this.storeMutation.onEndGame],
      [SlotGameEvent.BEGIN_GAME, this.storeMutation.onBeginGame],
      [SlotGameEvent.DOUBLE_GAME, this.storeMutation.onDoubleGame],
      [SlotGameEvent.KEEP_MACHINE_STATUS, this.storeMutation.onKeepMachineStatus]
    ].forEach(([eventName, handler]) => {
      useGlobalEventDispatcher().addEventListener(eventName, (param) => {
        if (param.event) {
          handler.bind(this.storeMutation)(param.data);
        }
      });
    });
  }
  insertStoreDataPattern(keyWithPattern) {
    return insertVariablePattern(
      this.storeGetter,
      keyWithPattern
    );
  }
  boot() {
    this.addListeners();
  }
}

export { CommonStore };
