// @gt-npm/gt-lib-ts/es/h5GameTools/commonStore/state.mjs
import '../../utils/reactivity/index.mjs';
import { reactive } from '../../utils/reactivity/vue-reactivity.mjs';

var StoreState;
((StoreState2) => {
  StoreState2.JPType = {
    Grand: 1,
    Minor: 2,
    Major: 3,
    Mini: 4
  };
})(StoreState || (StoreState = {}));
function useStoreState() {
  const baseData = reactive({
    userID: 0,
    line: 0,
    lineBet: 0,
    maxLine: 0,
    maxLineBet: 0,
    gameCode: "",
    machineList: [],
    exchangeCredit: 0,
    exchangeAmount: 0,
    jackpot: 0,
    marquee: "",
    winJPType: 0,
    winJPAmount: 0,
    loginName: "",
    balance: 0,
    credit: 0,
    creditEnd: 0,
    base: "",
    betBase: "",
    wagersID: "",
    encryID: "",
    levelID: 0,
    brickNum: 0,
    payoff: 0,
    bet: 0,
    cards: [],
    lines: "",
    scatter: "",
    bonus: "",
    free: "",
    freeTimes: 0,
    doubleTime: 0,
    axisLocation: "",
    betCreditList: [],
    defaultBetCredit: 0,
    isCash: false,
    enableBuyFreeGame: false,
    isGCP: false,
    isXC: parent.Site === "XC",
    defaultBase: "",
    noExchange: false,
    currency: "",
    i18n: {}
  });
  return baseData;
}

export { StoreState, useStoreState };
